toolsrc/src/vcpkg/versions.cpp

```diff
@@ -27,4 +46,138 @@ namespace vcpkg::Versions
 
         return hash<string>()(key.port_name) ^ (hash<string>()(key.version.to_string()) >> 1);
     }
+
+    SemanticVersion SemanticVersion::from_string(const std::string& str)
+    {
+        SemanticVersion ret;
+        ret.original_string = str;
+        ret.version_string = str;
+
+        auto build_found = ret.version_string.find('+');
+        if (build_found != std::string::npos)
+        {
+            ret.version_string.resize(build_found);
+        }
+
+        auto prerelease_found = ret.version_string.find('-');
+        if (prerelease_found != std::string::npos)
+        {
+            ret.prerelease_string = ret.version_string.substr(prerelease_found + 1);
+            ret.identifiers = std::move(Strings::split(ret.prerelease_string, '.'));
+            ret.version_string.resize(prerelease_found);
+        }
+
+        std::regex version_match("(0|[1-9][0-9]*)(\\.(0|[1-9][0-9]*)){2}");
+        Checks::check_exit(VCPKG_LINE_INFO,
+                           std::regex_match(ret.version_string, version_match),
+                           "Error: String `%s` is not a valid Semantic Version string.",
+                           str);
+
+        auto parts = Strings::split(ret.version_string, '.');
+        ret.version = std::move(
+            Util::fmap(parts, [](auto&& strval) -> long { return as_numeric(strval).value_or_exit(VCPKG_LINE_INFO); }));
+
+        return std::move(ret);
```

Comment:
```suggestion
        return ret;
```
`std::move()` is not required when returning a stack variable that exactly matches the return type -- it _is_ needed if doing a conversion from `T` to `Optional<T>`.